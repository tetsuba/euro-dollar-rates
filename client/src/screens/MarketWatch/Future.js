import React from 'react'
import service from '../../utils/service'
import { withRouter } from '../../HOC/withRouter'
import BreadCrumbs from '../../components/BreadCrumbs/BreadCrumbs'
import LineChart from '../../components/Charts/LineChart'
import {
    mutateFutureList,
    mutateFutureListForLineChart,
    mutateFutureStringToObject,
} from '../../utils/mutations'
import FutureCheckBoxList from '../../components/List/FutureCheckBoxList';

class Future extends React.Component {
    state = {
        future: {},
        showHideLineChart: false,
    }

    async componentDidMount() {
        const future = await service.getFuture(this.props.params.symbol)

        this.setState({
            future: future.data,
            futureList: mutateFutureList(future.data.list),
            futureYear: future.data.list.map(mutateFutureStringToObject),
        })
    }

    handlerOnClickCheckbox = (e) => {
        const { futureList } = this.state
        const index = e.target.getAttribute('index')
        const checked = e.target.getAttribute('data-checked')
        futureList[index].checked = !(checked === 'true')
        this.setState({
            futureList,
        })
    }

    handlerOnClickButton = () => {
        this.setState({
            showHideLineChart: !this.state.showHideLineChart,
        })
    }

    render() {
        const { showHideLineChart, futureList } = this.state

        return (
            <>
                <BreadCrumbs />
                {showHideLineChart && (
                    <>
                        <div className="row">
                            <button
                                type="button"
                                className="btn btn-primary"
                                onClick={this.handlerOnClickButton}
                            >
                                Back
                            </button>
                        </div>
                        <div className="row mt-5">
                            <LineChart
                                data={mutateFutureListForLineChart(futureList)}
                            />
                        </div>
                    </>
                )}
                {futureList && !showHideLineChart && (
                    <>
                        <div className="row">
                            <button
                                type="button"
                                className="btn btn-primary"
                                onClick={this.handlerOnClickButton}
                            >
                                Create Chart
                            </button>
                        </div>
                        <div className="row mt-5" data-testid="future-checkbox-list">
                            <FutureCheckBoxList
                              list={futureList}
                              onClick={this.handlerOnClickCheckbox}
                            />
                        </div>
                    </>
                )}
            </>
        )
    }
}

export default withRouter(Future)
