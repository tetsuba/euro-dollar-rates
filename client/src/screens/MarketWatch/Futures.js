import React from 'react'
import axios from 'axios'
import SearchBox from '../../components/SearchBox/SearchBox'
import FilterList from '../../components/List/FilterList'
import BreadCrumbs from '../../components/BreadCrumbs/BreadCrumbs'

const URLS = {
  MARKET_WATCH: 'https://www.marketwatch.com/investing/future/',
}

export default class Futures extends React.Component {
  state = {
    list: [],
    text: ''
  }

  async componentDidMount() {
    const futures = await axios.get('http://localhost:3000/api/futures/list')
    const list = futures.data.list.map(({name}) => name)
    this.setState({
      list,
      group: this.groupList(list)
    })
  }

  groupList(list) {
    const j = list
      .filter((string) => !string.endsWith('00'))
      .map((string) => string.replace(string.slice(-3), ''))

    return [...new Set(j)]
  }

  inputText = (ele) => {
    this.setState({
      text: ele.target.value.toUpperCase()
    })
  }

  render() {
    return (
      <>
        <BreadCrumbs />
        <div className="row">
          <div className="col">
            <SearchBox
              text={this.state.text}
              label="Futures Symbol"
              inputFunc={this.inputText}
            />
          </div>
        </div>
        <div className="row mt-4">
          <div className="col">
            <FilterList
              label="Full List"
              list={this.state.list}
              text={this.state.text}
              link={URLS.MARKET_WATCH}
            />
          </div>
          <div className="col">
            <FilterList
              label="Group List"
              list={this.state.group}
              text={this.state.text}
            />
          </div>
        </div>
      </>
    )
  }
}
