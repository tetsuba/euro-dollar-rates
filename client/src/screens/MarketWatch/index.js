import React from 'react'
import { Link } from 'react-router-dom'
import BreadCrumbs from '../../components/BreadCrumbs/BreadCrumbs'

export default class MarketWatch extends React.Component {
    render() {
        return (
            <>
                <BreadCrumbs />

                <h3>Market Watch</h3>
                <div className="row mt-3">
                    <div className="card" style={{ width: '18rem' }}>
                        <div className="card-body">
                            <h5 className="card-title">Futures</h5>
                            <p className="card-text">
                                Click on the button below to view market watch
                                futures list codes
                            </p>
                            <Link to="futures" className="btn btn-primary">
                                Click here
                            </Link>
                        </div>
                    </div>
                </div>
            </>
        )
    }
}
