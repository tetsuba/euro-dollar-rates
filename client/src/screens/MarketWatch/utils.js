

export function convertFutureStringToObject(future) {
  const [month, year] = future.name.slice(-8).split(' ')
  return {
    month,
    year,
    price: future.price
  }
}