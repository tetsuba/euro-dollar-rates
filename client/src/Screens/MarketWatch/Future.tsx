import React from 'react'
import service from '../../utils/service'

// HOC
import { withRouter } from '../../HOC/withRouter'

// COMPONENTS
import BreadCrumbs from '../../Components/BreadCrumbs/BreadCrumbs'
import LineChart from '../../Components/Charts/LineChart'
import FutureCheckBoxList from '../../Components/List/FutureCheckBoxList'

// UTILS
import {
    mutateFutureList,
    mutateFutureListForLineChart,
    mutateFutureStringToObject,
} from '../../utils/mutations'
import { EventClickType, FutureListType, FutureType } from '../../utils/types'
import LoadingSpinner from '../../Components/LoadingSpinner/LoadingSpinner'

interface PropTypes {
    params: {
        [key: string]: string
    }
}


interface StateTypes {
    future: {
        data: {
            list: FutureType[]
        }
    }
    futureList: FutureType[]
    futureYear: FutureListType[]
    showHideLineChart: boolean
}


class Future extends React.Component<PropTypes, StateTypes> {
    state = {
        future: {
            data: {
                list: [],
            },
        },
        futureList: [],
        futureYear: [],
        showHideLineChart: false,
    }

    async componentDidMount() {
        const future = await service.getFuture(this.props.params.symbol)

        this.setState({
            future: future.data,
            futureList: mutateFutureList(future.data.list),
            futureYear: future.data.list.map(mutateFutureStringToObject),
        })
    }

    handlerOnClickCheckbox = (e: EventClickType): void => {
        const { futureList } = this.state
        const el = e.target as HTMLElement
        const index: null | string = el.getAttribute('data-index')
        const checked: null | string = el.getAttribute('data-checked')

        // TODO: Look into this and find out why TS is producing this error
        // TS2339: Property 'checked' does not exist on type 'never'.
        // @ts-ignore
        futureList[Number(index)].checked = !(checked === 'true')
        // futureList[Number(index)].checked = !(checked === 'true')
        this.setState({ futureList })
    }

    handlerOnClickButton = () => {
        this.setState({
            showHideLineChart: !this.state.showHideLineChart,
        })
    }

    renderChart(futureList: FutureType[]) {
        return (
            <>
                <div className="row">
                    <button
                        type="button"
                        className="btn btn-primary"
                        onClick={this.handlerOnClickButton}
                    >
                        Back
                    </button>
                </div>
                <div className="row mt-5">
                    <LineChart {...mutateFutureListForLineChart(futureList)} />
                </div>
            </>
        )
    }

    renderList(futureList: FutureType[]) {
        return (
            <>
                <div className="row">
                    <button
                        type="button"
                        className="btn btn-primary"
                        onClick={this.handlerOnClickButton}
                    >
                        Create Chart
                    </button>
                </div>
                <div className="row mt-5" data-testid="future-checkbox-list">
                    <FutureCheckBoxList
                        list={futureList}
                        onClick={this.handlerOnClickCheckbox}
                    />
                </div>
            </>
        )
    }

    render() {
        const { showHideLineChart, futureList } = this.state

        return (
            <>
                <BreadCrumbs />
                {futureList.length < 1 ? (
                    <LoadingSpinner />
                ) : showHideLineChart ? (
                    this.renderChart(futureList)
                ) : (
                    this.renderList(futureList)
                )}
            </>
        )
    }
}

export default withRouter(Future)
