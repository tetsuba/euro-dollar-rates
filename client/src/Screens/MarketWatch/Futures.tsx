import React from 'react'
import { GET_FUTURES_LIST } from '../../utils/service'
import SearchBox from '../../Components/SearchBox/SearchBox'
import BreadCrumbs from '../../Components/BreadCrumbs/BreadCrumbs'
import FuturesFullList from '../../Components/List/FuturesFullList'
import FuturesGroupList from '../../Components/List/FuturesGroupList'
import withLoader from '../../HOC/withLoader'
import {
    mutateFuturesList,
    mutateFuturesListByGroupingNames,
} from '../../utils/mutations'
import { FuturesType, InputElementType } from '../../utils/types'

interface PropTypes {
    futures: FuturesType
}

interface StateTypes {
    list: string[]
    group: string[]
    text: string
}

class Futures extends React.Component<PropTypes, StateTypes> {
    state = {
        list: [],
        group: [],
        text: '',
    }

    async componentDidMount() {
        const list = mutateFuturesList(this.props.futures.list)
        this.setState({
            list,
            group: mutateFuturesListByGroupingNames(list),
        })
    }

    inputText = (e: InputElementType) => {
        this.setState({
            text: e.target.value.toUpperCase(),
        })
    }

    filterList(list: string[], text: string) {
        return list.filter((name: string) => name.startsWith(text))
    }

    render() {
        const { text, list, group } = this.state

        return (
            <>
                <BreadCrumbs />
                <div className="row">
                    <div className="col">
                        <SearchBox
                            text={text}
                            label="Futures Symbol"
                            inputFunc={this.inputText}
                        />
                    </div>
                </div>
                <div className="row mt-4">
                    <div className="col">
                        <FuturesFullList
                            list={this.filterList(list, text)}
                            text={text}
                        />
                    </div>
                    <div className="col">
                        <FuturesGroupList
                            list={this.filterList(group, text)}
                            text={text}
                        />
                    </div>
                </div>
            </>
        )
    }
}

// TODO: Using an HOC to grab data from the server.
//       This is not very clear to follow.
//       Look into using hooks pattern to handle this request.
export default withLoader(Futures, GET_FUTURES_LIST)
