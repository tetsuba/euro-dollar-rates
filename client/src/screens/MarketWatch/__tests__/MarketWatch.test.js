import { render } from '@testing-library/react'
import MarketWatch from '../MarketWatch'
import { MemoryRouter } from 'react-router-dom'

describe('MarketWatch', () => {
  test('should render screen', () => {
    const { asFragment } = render(<MemoryRouter><MarketWatch /></MemoryRouter>)
    expect(asFragment()).toMatchSnapshot()
  })
})