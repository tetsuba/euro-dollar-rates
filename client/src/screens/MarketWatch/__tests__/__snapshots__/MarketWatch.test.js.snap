// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarketWatch should render screen 1`] = `
<DocumentFragment>
  <div
    class="row shadow mb-5"
  >
    <nav
      aria-label="breadcrumb"
    >
      <ol
        class="breadcrumb mt-3"
      >
        <li
          class="breadcrumb-item"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          aria-current="page"
          class="breadcrumb-item active"
        />
      </ol>
    </nav>
  </div>
  <h3>
    Market Watch
  </h3>
  <div
    class="row mt-3"
  >
    <div
      class="card"
      style="width: 18rem;"
    >
      <div
        class="card-body"
      >
        <h5
          class="card-title"
        >
          Futures
        </h5>
        <p
          class="card-text"
        >
          Click on the button below to view market watch futures list codes
        </p>
        <a
          class="btn btn-primary"
          href="/futures"
        >
          Click here
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
